In [38]:
import pandas as pd
import numpy as np

import os
import glob

import matplotlib.pyplot as plt
from IPython.display import display

# Data exploration

In [19]:
Path = r"C:\Users\Samsung\Desktop\all\Project\23_VRPTW\Data\aws\amazon_data\almrrc2021-data-training"

In [28]:
os.chdir(Path)
file_list = os.listdir()
file_all = []
for fi_ in file_list:
    path_ = Path + "\\" + fi_
    os.chdir(path_)
    target_ex = "*.json"
    if len(glob.glob(target_ex)) != 0:
        for fi_2 in glob.glob(target_ex):
            file_all.append([fi_2,path_])
    else:
        pass

In [30]:
json_list = [i[0] for i in file_all]
json_list

['new_package_data.json',
 'new_route_data.json',
 'new_travel_times.json',
 'proposed_sequences.json',
 'actual_sequences.json',
 'invalid_sequence_scores.json',
 'package_data.json',
 'route_data.json',
 'travel_times.json',
 'new_actual_sequences.json',
 'new_invalid_sequence_scores.json',
 'scores.json',
 'model_apply_time.json',
 'model_build_time.json']

### Check name duplication

In [31]:
len(json_list)

14

In [32]:
len(set(json_list))

14

All json file name is uniqure regardless of files

## loading data (packages data)

In [43]:
# model_build_inputs / package_data
path_ = r"C:\Users\Samsung\Desktop\all\Project\23_VRPTW\Data\aws\amazon_data\almrrc2021-data-training\model_build_inputs\package_data.json"
package_data = pd.read_json(path_)

In [101]:
package_data = package_data.fillna("0")

In [102]:
package_data.head()

RouteID_00143bdd-0a6b-49ec-bb35-36593d303e77  \
AD  {'PackageID_9d7fdd03-f2cf-4c6f-9128-028258fc09...   
AF  {'PackageID_15c6a204-ec5f-4ced-9c3d-472316cc77...   
AG  {'PackageID_3b28f781-242e-416e-9575-84c7188b82...   
BA  {'PackageID_a18e36e0-6b5a-45b7-bf7a-7c3a15bd5a...   
BE  {'PackageID_22c8f630-8ab9-40d5-a36d-38ef01761b...   

         RouteID_0016bc70-cb8d-48b0-aa55-8ee50bdcdb59  \
AD                                                  0   
AF                                                  0   
AG  {'PackageID_62bd35ae-6fe9-40c0-8c58-0d1021a796...   
BA  {'PackageID_80d944eb-442a-422f-9d40-3ad4bb73d7...   
BE                                                  0   

         RouteID_001948e9-4675-486d-9ec5-912fd8e0770f  \
AD  {'PackageID_de2b05b8-2186-47b2-b0f1-640cd01b65...   
AF                                                  0   
AG                                                  0   
BA                                                  0   
BE                                                  0   

         RouteID_001b4ee3-c4f2-467f-932b-c85524d1021f  \
AD                                                  0   
AF                                                  0   
AG  {'PackageID_1520728c-7a38-45a1-9e5d-123eb6d023...   
BA                                                  0   
BE  {'PackageID_d3c08539-cea1-4079-a0a2-f064fc2083...   

         RouteID_0021a2aa-780f-460d-b09a-f301709e2523  \
AD                                                  0   
AF  {'PackageID_ef2e9ad8-43a7-4043-9144-faf6730da3...   
AG  {'PackageID_9f9e7677-9f0f-42c4-ac05-8015c1447d...   
BA                                                  0   
BE                                                  0   

         RouteID_0036cf37-1c93-48f8-b337-3dd6d1a195d8  \
AD                                                  0   
AF                                                  0   
AG                                                  0   
BA                                                  0   
BE  {'PackageID_076224e5-3e6f-4b0a-b927-b74c185412...   

   RouteID_00437423-5a8f-4fac-8efa-85264f0a3944  \
AD                                            0   
AF                                            0   
AG                                            0   
BA                                            0   
BE                                            0   

         RouteID_00573a3d-b909-42ee-8a99-c6474309d96b  \
AD                                                  0   
AF                                                  0   
AG  {'PackageID_66dffd90-5ed1-467a-bb08-29a6a75b13...   
BA                                                  0   
BE  {'PackageID_d7f87b82-1d83-413e-a3d6-4dc2f6ede7...   

         RouteID_00575a14-7d19-4291-8853-8f7095ed1bc6  \
AD  {'PackageID_2567ffe2-6172-4730-951f-4eeb477dc7...   
AF                                                  0   
AG  {'PackageID_516e130d-177c-4323-a25e-d00b61b3ce...   
BA                                                  0   
BE                                                  0   

         RouteID_00575ca4-8a63-49d2-96c8-9b347be5ba6c  ...  \
AD                                                  0  ...   
AF                                                  0  ...   
AG  {'PackageID_f0f34a05-5434-4a6a-ba53-fa7b814e15...  ...   
BA                                                  0  ...   
BE                                                  0  ...   

         RouteID_ff709338-da35-4864-ab32-6bcd8b75f953  \
AD  {'PackageID_6da1aa31-1d40-453e-af28-28c3f895f5...   
AF                                                  0   
AG  {'PackageID_695dd622-6a8c-4e88-b26d-8254d0617e...   
BA  {'PackageID_8d087e79-11d3-4b84-920d-8016029333...   
BE                                                  0   

         RouteID_ff915688-4784-49a0-9ee6-f424f1768ca4  \
AD                                                  0   
AF                                                  0   
AG  {'PackageID_7d68e440-b27b-4b75-9f67-0b01dad51b...   
BA            

### Counting missing value

In [115]:
count_ = 0
for index,col in package_data.iterrows():
    count_ += len([i for i in col if i == "0"])
    

In [116]:
count_

3221073

In [119]:
total_ = len(package_data.index) * len(package_data.columns)
total_

4125600

In [121]:
(count_/total_)* 100

78.07526178010471

Abount Missing value

Total = 4125600
Missing vlaue = 3221073
Missing persentages = 78%

In [103]:
len(package_data.index)

675

In [104]:
len(package_data.columns)

6112

Data featuring

Number of Stop location: 675
Number of Route : 6112

In [59]:
package_data.iloc[0,0]

{'PackageID_9d7fdd03-f2cf-4c6f-9128-028258fc09ea': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': None, 'end_time_utc': None},
  'planned_service_time_seconds': 59.3,
  'dimensions': {'depth_cm': 25.4, 'height_cm': 7.6, 'width_cm': 17.8}},
 'PackageID_5541e679-b7bd-4992-b288-e862f6c84ae7': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': '2018-07-27 16:00:00',
   'end_time_utc': '2018-07-28 00:00:00'},
  'planned_service_time_seconds': 59.3,
  'dimensions': {'depth_cm': 25.4, 'height_cm': 12.7, 'width_cm': 17.8}},
 'PackageID_84d0295b-1adb-4a33-a65e-f7d6247c7a07': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': None, 'end_time_utc': None},
  'planned_service_time_seconds': 59.3,
  'dimensions': {'depth_cm': 39.4, 'height_cm': 7.6, 'width_cm': 31.8}}}

In [ ]:
# Check whether the packageID is an 

### Extracting packageID data removing stop-id, route-id
--> 다시 위에 데이터 까지 포함시켜서 (특히 Location)

In [158]:
df_1 = [col_ for index_,col_ in package_data.T.iterrows()]
#df_onlypack = df_1

In [134]:
[i for i in df_1[0] if i != "0"][1]

{'PackageID_15c6a204-ec5f-4ced-9c3d-472316cc7759': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': '2018-07-27 16:00:00',
   'end_time_utc': '2018-07-28 00:00:00'},
  'planned_service_time_seconds': 27.0,
  'dimensions': {'depth_cm': 30.0, 'height_cm': 3.0, 'width_cm': 27.4}}}

In [163]:
df_1[1]

AD                                                    0
AF                                                    0
AG    {'PackageID_62bd35ae-6fe9-40c0-8c58-0d1021a796...
BA    {'PackageID_80d944eb-442a-422f-9d40-3ad4bb73d7...
BE                                                    0
                            ...                        
SX                                                    0
SU                                                    0
UK                                                    0
VK                                                    0
BN                                                    0
Name: RouteID_0016bc70-cb8d-48b0-aa55-8ee50bdcdb59, Length: 675, dtype: object

In [164]:
dic_all = {}
for dic_ in df_1:
    dic_can = [i for i in dic_ if i != "0"]
    print(dic_can)
    break
    dic_all.update(dic_can)

[{'PackageID_9d7fdd03-f2cf-4c6f-9128-028258fc09ea': {'scan_status': 'DELIVERED', 'time_window': {'start_time_utc': None, 'end_time_utc': None}, 'planned_service_time_seconds': 59.3, 'dimensions': {'depth_cm': 25.4, 'height_cm': 7.6, 'width_cm': 17.8}}, 'PackageID_5541e679-b7bd-4992-b288-e862f6c84ae7': {'scan_status': 'DELIVERED', 'time_window': {'start_time_utc': '2018-07-27 16:00:00', 'end_time_utc': '2018-07-28 00:00:00'}, 'planned_service_time_seconds': 59.3, 'dimensions': {'depth_cm': 25.4, 'height_cm': 12.7, 'width_cm': 17.8}}, 'PackageID_84d0295b-1adb-4a33-a65e-f7d6247c7a07': {'scan_status': 'DELIVERED', 'time_window': {'start_time_utc': None, 'end_time_utc': None}, 'planned_service_time_seconds': 59.3, 'dimensions': {'depth_cm': 39.4, 'height_cm': 7.6, 'width_cm': 31.8}}}, {'PackageID_15c6a204-ec5f-4ced-9c3d-472316cc7759': {'scan_status': 'DELIVERED', 'time_window': {'start_time_utc': '2018-07-27 16:00:00', 'end_time_utc': '2018-07-28 00:00:00'}, 'planned_service_time_seconds': 

### Q) check whether packageID is duplicated or an unique value

--> No are not shared between routes

In [141]:
len(dic_all.keys())

10158

In [142]:
len(set(dic_all.keys()))

10158

In [ ]:
# all inidividual 

In [146]:
dic_all

{'PackageID_15c6a204-ec5f-4ced-9c3d-472316cc7759': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': '2018-07-27 16:00:00',
   'end_time_utc': '2018-07-28 00:00:00'},
  'planned_service_time_seconds': 27.0,
  'dimensions': {'depth_cm': 30.0, 'height_cm': 3.0, 'width_cm': 27.4}},
 'PackageID_80d944eb-442a-422f-9d40-3ad4bb73d73c': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': None, 'end_time_utc': None},
  'planned_service_time_seconds': 36.3,
  'dimensions': {'depth_cm': 25.4, 'height_cm': 12.7, 'width_cm': 17.8}},
 'PackageID_fb1f8f8d-7bb1-47c4-b149-bb99e49cfa24': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': None, 'end_time_utc': None},
  'planned_service_time_seconds': 36.3,
  'dimensions': {'depth_cm': 43.2, 'height_cm': 7.6, 'width_cm': 27.9}},
 'PackageID_0f21ecd7-a584-4e35-bdf1-6a10cd2923d3': {'scan_status': 'DELIVERED',
  'time_window': {'start_time_utc': None, 'end_time_utc': None},
  'planned_service_time_seconds': 36.3,
  '

In [145]:
## are not sare shared between routes